### ISSUE TITLE
Improve keyboard mapping UI

### ISSUE THREAD

**Issue by jtsiomb** [FreeSpacenav/spnavcfg#35]
Instead of having a list of a handful common keysyms, and allowing the user to type keysym names in the field, it would be much simpler to have a button which asks the user to press the key they want to map.

**Comment by TimSchimansky** [comment]
+1, that would be great. Basically making the buttons a mini macro pad

**Comment by TyanColte** [comment]
yes, this would be great in OrcaSlicer where i can hit Ctrl+0 to reset the view.

**Comment by jtsiomb** [comment]
Multi-key mappings are already supported by spacenavd. It's just the spnavcfg UI that hasn't caught up. You can configure that from the config file directly.

**Comment by TyanColte** [comment]
> Multi-key mappings are already supported by spacenavd. It's just the spnavcfg UI that hasn't caught up. You can configure that from the config file directly.

What if I don't have a config file?

**Comment by jtsiomb** [comment]
If you don't have a config file, you're using all the options at their defaults. A config file is created when you hit save from spnavcfg, or you can just copy the `example-spnavrc` file, from `doc/` in the spacenavd source tree, as `/etc/spnavrc` and modify it.

**Comment by TyanColte** [comment]
I've hit save from spnavcfg but there is still no config file in /etc/spnavrc. i wonder if it has something to do with me being on the immutable distro Bazzite.

**Comment by jtsiomb** [comment]
I'm not familiar with your distro, but immutable doesn't sound promising :) Refer to your distro documentation about how to create config files in /etc. Maybe there's a process to temporarily mount root read-write to make changes.

Alternatively, run spacenavd with the `-c` option to specify a different configuration file path, to a location where you have access. See `spacenavd -h` for all the available command-line options.

**Comment by TyanColte** [comment]
well it's definitely putting a configuration file somewhere otherwise the save wouldn't be successful. Nor would it retain it's settings upon closing and reopening the app, is there a way to see verbose stdout while the app is running from terminal to see if it logs a save location?

**Comment by jtsiomb** [comment]
When spacenavd starts, it logs the location of the config file, so you can look there. The default location of the log file is `/var/log/spnavd.log`, but that too can be changed by command-line argument.

But it's just going to be whatever you passed to `-c` or `/etc/spnavrc` by default.

    $ cat /var/log/spnavd.log | grep config
    reading config file: /etc/spnavrc

**Comment by TyanColte** [comment]
I found it, right where it should've been. Not sure why I couldn't find it before. But now even based on the example config, I can't figure out how to set button 1 to do Ctrl+0 on my keyboard. (That's the reset view command on Orca Slicer) Any suggestions?

**Comment by jtsiomb** [comment]
Something like `kbmap0 = Control_L+0`.
In general look into `/usr/include/X11/keysymdef.h` for the keysym names, and disregard the `XK_` prefix.

**Comment by krenslow** [comment]
I have just discovered a new way to binding keys to the Spacemouse Enterprise that allows for more than just simple key combos. I'm not sure if this works outside Ubuntu. I used the spnavcfg program and mapped every button to any key such that every button has a unique key. Doesn't matter what key. Then I installed input-remapper through apt and created a profile on `spacenavd virtual keyboard`. I add an input, push record, hit the button on mouse, and then it will give the option to remap the button to a "key or macro". The macros on input-remapper allows for delays, if blocks, variables, chaining keys, etc.
